pkg/nitrite/nitrite.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/blocky/nitrite/pkg/cose"
+)
+
+// Result is a successful verification result of an attestation payload.
+type Result struct {
+	// Document contains the attestation document.
+	Document *attestation.Document `json:"document,omitempty"`
+
+	// Certificates contains all of the certificates except the root.
+	Certificates []*x509.Certificate `json:"certificates,omitempty"`
+
+	// Protected section from the COSE Sign1 payload.
+	Protected []byte `json:"protected,omitempty"`
+	// Unprotected section from the COSE Sign1 payload.
+	Unprotected []byte `json:"unprotected,omitempty"`
+	// Payload section from the COSE Sign1 payload.
+	Payload []byte `json:"payload,omitempty"`
+	// Signature section from the COSE Sign1 payload.
+	Signature []byte `json:"signature,omitempty"`
+
+	// SignatureOK designates if the signature was OK (but certificate could be
+	// invalid, not trusted, expired, etc.)
+	SignatureOK bool `json:"signature_ok"`
+
+	// COSESign1 contains the COSE Signature Structure which was used to
+	// calculate the `Signature`.
+	COSESign1 []byte `json:"cose_sign1,omitempty"`
+}
+
+// VerifyOptions specifies the options for verifying the attestation payload.
+// If `Roots` is nil, the `DefaultCARoot` is used. If `CurrentTime` is 0,
+// `time.Now()` will be used. It is a strong recommendation you explicitly
+// supply this value.
+type VerifyOptions struct {
+	Roots               *x509.CertPool
+	CurrentTime         time.Time
+	AllowSelfSignedCert bool
+}
+
+// Verify verifies the attestation payload from `data` with the provided
+// verification options. If the options specify `Roots` as `nil`, the
+// `DefaultCARoot` will be used. If you do not specify `CurrentTime`,
+// `time.Now()` will be used. It is strongly recommended you specifically
+// supply the time.  If the returned error is non-nil, it is either one of the
+// `Err` codes specified in this package, or is an error from the `crypto/x509`
+// package. Revocation checks are NOT performed and you should check for
+// revoked certificates by looking at the `Certificates` field in the `Result`.
+// Result will be non-null if and only if either of these are true: certificate
+// verification has passed, certificate verification has failed (expired, not
+// trusted, etc.), signature is OK or signature is not OK. If either signature
+// is not OK or certificate can't be verified, both Result and error will be
+// set! You can use the SignatureOK field from the result to distinguish
+// errors.
+func Verify(data []byte, options VerifyOptions) (*Result, error) {
+	cosePayload, err := cose.ExtractCosePayload(data)
+	if err != nil {
+		return nil, err
+	}
+	err = cose.VerifyCosePayload(cosePayload)
+	if err != nil {
+		return nil, err
+	}
+
+	coseHeader, err := cose.ExtractCoseHeader(cosePayload)
+	if err != nil {
+		return nil, err
+	}
+	err = cose.VerifyCoseHeader(coseHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err := cose.ExtractAttestationDoc(cosePayload)
```

Comment:
Here you could call `cbor.Unmarshal(payload.Payload, &doc)` to get the `doc`. The main reason is that while a `COSEPayload` should know how to verify itself and ask its components to verify themselves, it does not need to know that the `Payload` field contains an `attestation.Document`. As a result `ExtractAttestationDoc` doesn't really belong in the `cose` package.